src/_pytest/_code/code.py

```diff
@@ -572,6 +572,11 @@ def match(self, regexp):
         raised.
         """
         __tracebackhide__ = True
+        if isinstance(regexp, text_type):              
+            if not re.search(regexp, text_type(self.value)):
+                assert 0, u"Pattern '{!s}' not found in '{!s}'".format(regexp, self.value)
+            return True
+
         if not re.search(regexp, str(self.value)):
```

Comment:
Actually I think the only change required is to use `text_type(self.value)` instead of `str(self.value)` here.

Reply:
This doesn't work when you have a non-ascii bytes exception message

Reply:
Oh you are absolutely right, `self.value` might be already encoded. So to avoid duplication I recommend to change to this:

```python
        __tracebackhide__ = True
        value = text_type(self.value) if isinstance(regexp, text_type) else str(self.value)
        if not re.search(regexp, value):
            assert 0, "Pattern '{!s}' not found in '{!s}'".format(regexp, value)
        return True
```

Reply:
But then the formatting won't be safe

Reply:
You are right, but I think this solves it:

```python
assert 0, u"Pattern '{!s}' not found in '{!s}'".format(regexp, value)
```

It should be safe to always use unicode messages with the assert statement

Reply:
or you could use a repr instead?
`assert 0, u"Pattern {!r} not found in {!r}".format(regexp, value)`

Reply:
how about safe_str ? or one of the to ascii helper s we have in the python2 support code?